src/ncp-spinel/SpinelNCPInstance.cpp

```diff
@@ -654,25 +644,30 @@ SpinelNCPInstance::update_node_type(NodeType new_node_type)
 	}
 }
 
-
 void
 SpinelNCPInstance::update_link_local_address(struct in6_addr *addr)
 {
 	if (NULL != addr
 	  && (0 != memcmp(mNCPLinkLocalAddress.s6_addr, addr->s6_addr, sizeof(mNCPLinkLocalAddress)))
 	) {
 		if (IN6_IS_ADDR_LINKLOCAL(&mNCPLinkLocalAddress)) {
-			remove_address(mNCPLinkLocalAddress);
+			unicast_address_was_removed(kOriginThreadNCP, mNCPLinkLocalAddress);
 		}
 
 		memcpy((void*)mNCPLinkLocalAddress.s6_addr, (void*)addr->s6_addr, sizeof(mNCPLinkLocalAddress));
 
-		if (IN6_IS_ADDR_LINKLOCAL(&mNCPLinkLocalAddress)) {
-			add_address(mNCPLinkLocalAddress);
-		}
-
 		signal_property_changed(kWPANTUNDProperty_IPv6LinkLocalAddress, in6_addr_to_string(*addr));
 	}
+
+	// Link-local address is always re-added to ensure that it is present after
+	// a reset where the address table maintained by `NCPInstanceBase` is wiped
+	// but the `mNCPLinkLocalAddress` remains valid. Note that we don't want to
+	// clear `mNcpLinkLocalAddress` so not to emit a property changed signal for
+	// `kWPANTUNDProperty_IPv6LinkLocalAddress` for an NCP reset.
+
+	if (IN6_IS_ADDR_LINKLOCAL(&mNCPLinkLocalAddress)) {
+		unicast_address_was_added(kOriginThreadNCP, mNCPLinkLocalAddress);
```

Comment:
With new model, all addresses are being tracked by `NCPInstanceBase` and  if we already had the address in our table, no action will be taken. If it's a new one then we ensure to sync/push it on other interface (i.e., if NCP added it we add it on host interface and vice versa)... So technically the same address would/should not be added multiple times on the primary interface...

From plug-in sub-class we always notify our base class that an addess was added/rmoved on NCP. The base class ensures to keep track of the state in its table.